lib/semantics/check-omp-structure.cc

```diff
@@ -184,9 +199,17 @@ void OmpStructureChecker::Enter(const parser::OmpClause::Notinbranch &) {
 void OmpStructureChecker::Enter(const parser::OmpClause::Untied &) {
   CheckAllowed(OmpClause::UNTIED);
 }
-void OmpStructureChecker::Enter(const parser::OmpClause::Collapse &) {
+
+void OmpStructureChecker::Enter(const parser::OmpClause::Collapse &x) {
   CheckAllowed(OmpClause::COLLAPSE);
+  // collapse clause must have a parameter
+  if (!GetPosIntValue(x.v)) {
+    context_.Say(GetContext().clauseSource,
+        "The parameter of the COLLAPSE clause must be "
+        "a constant positive integer expression"_err_en_US);
```

Comment:
I think you'll get two errors here if the expression is not constant or not integer. It would be better just to have one.

Reply:
You're right, I don't have a good solution here.
```
do.f90:3:12: error: The parameter of the COLLAPSE clause must be a constant positive integer expression
    !$omp do collapse(1.0) ordered(1+2)
             ^^^^^^^^^^^^^
do.f90:3:21: error: Must have INTEGER type, but is REAL(4)
    !$omp do collapse(1.0) ordered(1+2)
                      ^^^
f18: semantic errors in do.f90
```

Reply:
You should be able to do it like for NUM_THREADS:
```
  const auto v{GetIntValue(x.v)};
  if (v && v.value() <= 0) {
```
`GetIntValue` calls `GetExpr` which should enforce the constraints that are expressed in the parse tree, `ScalarIntConstantExpr` in this case. All you have to check is that it is positive.

Reply:
Addressed in commit [6ad5f9f](https://github.com/flang-compiler/f18/pull/540/commits/6ad5f9f4427af9618137fa8036519c256d2303dc). Non-constant integer expr will be deferred to Expr Analysis. Thanks for the advice.